File: packages/quran-engine-swift/Sources/QuranEngine/Cache.swift
import Foundation

/// Sanitize a reciter id into a filesystem-safe directory name. Mirrors `sanitizeReciterDir`:
/// keep `[A-Za-z0-9-_]`, replace everything else with `_`, cap at 180 chars, fallback `"reciter"`.
public func sanitizeReciterDir(_ reciterId: String) -> String {
    var chars = [Character]()
    for ch in reciterId {
        if ch.isASCII && (ch.isLetter || ch.isNumber || ch == "-" || ch == "_") {
            chars.append(ch)
        } else {
            chars.append("_")
        }
        if chars.count == 180 { break }
    }
    let safe = String(chars)
    return safe.isEmpty ? "reciter" : safe
}

/// Relative path (under the downloads root) for a downloaded full-surah file.
/// Mirrors `localSurahPath`.
public func localSurahPath(_ reciter: Reciter, _ surahNumber: Int) -> String {
    "\(sanitizeReciterDir(reciter.id))/\(zeroPad3(surahNumber)).mp3"
}

/// Relative path of the content-addressed shared file for a given content hash.
public func sharedAudioPath(_ sha256Hex: String, ext: String = "mp3") -> String {
    "SharedAudio/\(sha256Hex).\(ext)"
}

File: packages/quran-engine-rust/src/cache.rs
//! Caching path helpers. Mirrors the path-builder portion of `src/cache.js`.
//!
//! The engine stays storage-agnostic — it only produces the canonical cache paths/keys.
//! Layout: `<root>/<sanitize(reciter.id)>/<zeroPad3(surah)>.mp3`.

use crate::util::zero_pad3;

/// Sanitize a reciter id into a filesystem-safe directory name: keep `[A-Za-z0-9-_]`,
/// replace everything else with `_`, cap at 180 chars (fallback `"reciter"`).
pub fn sanitize_reciter_dir(reciter_id: &str) -> String {
    let mut safe: String = reciter_id
        .chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || c == '-' || c == '_' {
                c
            } else {
                '_'
            }
        })
        .take(180)
        .collect();
    if safe.is_empty() {
        safe = "reciter".to_string();
    }
    safe
}

/// Relative path (under the downloads root) for a downloaded full-surah file:
/// `sanitize(reciter.id)/zeroPad3(surah).mp3`.
pub fn local_surah_path(reciter_id: &str, surah: u32) -> String {
    format!("{}/{}.mp3", sanitize_reciter_dir(reciter_id), zero_pad3(surah))
}

/// Relative path of the content-addressed shared file for a given content hash.
pub fn shared_audio_path(sha256_hex: &str, ext: &str) -> String {
    format!("SharedAudio/{}.{}", sha256_hex, ext)
}
